custom_components/adaptive_lighting/switch.py

```diff
@@ -1127,6 +1210,74 @@ async def turn_on(service_data):
                 await asyncio.sleep(self._send_split_delay / 1000.0)
                 await turn_on(service_datas[1])
 
+    # Not precise, checked every `interval`
+    async def _maybe_reset_manual_control(
+        self,
+        now: list[str] | None = None,
+    ):
+        for light in self._lights:
+            if self._manual_lights[light]["timer"] == 0:
+                continue
+            time_elapsed = perf_counter() - self._manual_lights[light]["timer"]
+            time_remaining = self._autoreset_control_time - time_elapsed
+            _LOGGER.debug(
+                "check manual control reset. time: %s, timer: %s elapsed: %s, remaining: %s",
+                self._autoreset_control_time,
+                self._manual_lights[light]["timer"],
+                time_elapsed,
+                time_remaining,
+            )
+            if time_remaining <= 0:
+                _LOGGER.debug("DISABLING manual control for lights %s", self._lights)
+                self.turn_on_off_listener.reset(*self._lights)
+
+    # Fixes #447 (and #430 with 'alt_detect_method: true' set)
+    # if we are in the middle of a transition, sleep until that transition finishes.
+    # Needed so we can check if the previous adapt attempt succeeded after the
+    # transition finalizes the new state.
+    # Someone with a better understanding of async usage could clean this up, not
+    # sure if it's safe to check a variable in another thread.
+    # The '_transitioning' variable is simply set to prevent buildups of multiple
+    # calls to this function all running at once after a long transition finishes
+    # with a short interval set.
+    async def _async_wait_transitions(self, now=None) -> bool:
+        _LOGGER.debug("%s: Check if we need to wait for transitions.", self._name)
+        if self._transitioning:
+            _LOGGER.debug(
+                "%s: Waiting for last transition to finish before attempting another adapt.",
+                self._name,
+            )
+            return True
+        _LOGGER.debug("%s: First call of _async_wait_transitions", self._name)
+        last_service_data = self.turn_on_off_listener.last_service_data
+        _LOGGER.debug("%s: Last service data: %s", self._name, last_service_data)
+        if (
+            not last_service_data
+            or not len(self._lights)
+            or self._lights[0] not in last_service_data
+            or ATTR_TRANSITION not in last_service_data[self._lights[0]]
+        ):
+            return False
+        last_transition = last_service_data[self._lights[0]][ATTR_TRANSITION]
+        _LOGGER.debug(
+            "%s: wait_transitions: Transition found in last service"
+            "data: %s seconds Timer: %s. Time remaining: %s",
+            self._name,
+            last_transition,
+            self._transition_timer,
+            (last_transition - (perf_counter() - self._transition_timer)),
+        )
+        while (
```

Comment:
One shouldn't do busy while loops in async code because it will create high CPU work.

Why don't you `await asyncio.sleep` the right amount of time?